### ISSUE TITLE
AutoStartable components not starting in time!

### ISSUE THREAD

**Issue by Infiniverse** [restx/restx#208]
My Restx app start just fine in dev mode, but it's exploding on a null pointer if I try and start it in 'prod' mode.

The reason stems from the RestxMainRouter: 

    private RestxMainRouter build(final String serverId, Optional<String> baseUri) {
        ...

        if (getLoadFactoryMode().equals("onstartup")) {

            Factory factory = loadFactory(newFactoryBuilder(serverId));
            factory = Factory.register(serverId, factory);

            StdRestxMainRouter mainRouter = newStdRouter(factory);
            routers.put(serverId, mainRouter);
            factory.start().and().prepare();

            logPrompt(baseUri, "READY", mainRouter);

            return mainRouter;
        }
        ....
     }

The main router is constructed with newStdRouter() which creates all the components. Only once everything is constructed does the factory.start() which starts any AutoStartable components.

This is bound to fail, if the AutoStartable exposes methods that are used in the constructers of other components, but which don't work until the component has been started.

For example:

@Component @RestxResource
public class MyResource {
        MyResource(B theB) {}
}

@Component
public class A implements AutoStartable {
    public boolean started = false;

    @Override
    public void start() {
        started = true;
    }
}

@Component
public class B {
    B(A theA) {
        if (!theA.started) {
            throw new RuntimeException("A not started!");
        }
    }
}

This fails in prod mode because A hasn't been started yet.

In my case A is a singleton SpringContext and I've got a module that provides access to it's beans as RestX components. However none of those components are valid unless the SpringContext has been started, so I'm getting null pointers at wiring time all over the place :(.

Connected to the target VM, address: '127.0.0.1:60932', transport: 'socket'
2015-06-25 07:41:59,472 [main            ] [          ] INFO  org.eclipse.jetty.server.Server - jetty-8.1.8.v20121106
2015-06-25 07:42:02,553 [main            ] [          ] INFO  restx.RestxMainRouterFactory - LOADING MAIN ROUTER
2015-06-25 07:42:05,073 [main            ] [          ] WARN  / - unavailable
java.lang.RuntimeException: A not started!
	at c.t.t.restx.B.<init>(B.java:9) ~[classes/:na]
	at c.t.t.restx.BFactoryMachine$1.doNewComponent(BFactoryMachine.java:24) ~[classes/:na]
	at c.t.t.restx.BFactoryMachine$1.doNewComponent(BFactoryMachine.java:12) ~[classes/:na]
	at restx.factory.StdMachineEngine.newComponent(StdMachineEngine.java:37) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory.buildAndStore(Factory.java:1219) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory.buildAndStore(Factory.java:1205) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory.satisfyBoms(Factory.java:1175) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory.buildAndStore(Factory.java:1069) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory.access$600(Factory.java:37) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory$NameQuery.doFindOne(Factory.java:604) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory$SingleQuery.doFind(Factory.java:521) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory$Query.find(Factory.java:448) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory$ClassQuery.doFind(Factory.java:662) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory$MultipleQuery.doFindOne(Factory.java:486) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory$Query.findOne(Factory.java:437) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory$Query.findOneAsComponent(Factory.java:440) ~[restx-factory-0.34.jar:na]
	at restx.factory.Factory.getComponent(Factory.java:798) ~[restx-factory-0.34.jar:na]
	at restx.RestxMainRouterFactory.newStdRouter(RestxMainRouterFactory.java:544) ~[restx-core-0.34.jar:na]
	at restx.RestxMainRouterFactory.build(RestxMainRouterFactory.java:375) ~[restx-core-0.34.jar:na]
	at restx.RestxMainRouterFactory.newInstance(RestxMainRouterFactory.java:70) ~[restx-core-0.34.jar:na]
	at restx.servlet.RestxMainRouterServlet.init(RestxMainRouterServlet.java:74) ~[restx-servlet-0.34.jar:na]

Shouldn't AutoStartable components be started as soon as they are constructed, rather than deferring start until a later phase? I can't see any reason why that ought to be a problem, and it would enable a wider variety of use cases than are currently supported (as well as reduce support noise around the use of AutoStartable components! :) )

Thanks,
Joe

**Comment by a-peyrard** [comment]
I don't have the answer why the "start and prepare" is done after the router initialisation, but I don't get why its a problem for you either :)

Why did B need a started A ? (I'm struggling to find a case where it would be a problem, maybe with something more concrete, it would be simple to seize the problem).
Cause I guess with another design you wouldn't have the problem, maybe with some lazy access to A from B and not direct access from the constructor. Maybe B should be the autostartable component, and call A initialisation... Once again I need more detail about B and A to figure out why it is a real issue.

**Comment by fcamblor** [comment]
:+1: with Augustin here

I suppose this is at least simpler to instantiate & inject every component in order to check if every BOM are satisfied, then start every components (in any case, it would have been the natural way I would have implemented things)

**Comment by Infiniverse** [comment]
I've been discussing this at length on the mailing list. My problem arises because some of my components are only available when my database is online. I have a factory component which has methods that return other components (through a module that calls its methods), but those methods only make sense once the component is started.

I agree that I can redesign and use lazy access, but I've got a lot of pre-existing components which I'm not able to change yet. (The app was a spring app which I've been migrating to restx - unfortunately the spring data component can't be initialised outside of spring and so I'm stuck with some beans that need to be exposed as components).

I've found a resolution, which is to make my spring bean factory AutoStartable, but to start it in the constructor instead of the start() method.

I now don't think that restx needs changing, but the documentation around AutoStartable needs to be more explicit about when the start occurs with respect to the wiring of dependencies. One would naturally expect a component to be started before it is used, but this isn't clear.

**Comment by Infiniverse** [comment]
I've pushed a suggested change to the web site docs. (https://github.com/restx/restx.github.io/pull/12)

**Comment by fcamblor** [comment]
```the documentation around AutoStartable needs to be more explicit about when the start occurs with respect to the wiring of dependencies. One would naturally expect a component to be started before it is used, but this isn't clear.```

Agreed, merged PR
